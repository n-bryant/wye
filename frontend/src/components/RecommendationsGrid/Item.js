import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";
import { withStyles } from "@material-ui/core/styles";

import Link from "next/link";
import Icon from "@mdi/react";
import { mdiThumbUpOutline, mdiThumbDownOutline } from "@mdi/js";

import Grid from "@material-ui/core/Grid";
import Paper from "@material-ui/core/Paper";
import Typography from "@material-ui/core/Typography";

import formatCurrency from "../../../lib/formatCurrency";
import HeroImage from "../Game/HeroImage";
import styles from "./Item.styles";

/**
 * renders a grid item holding data for a game
 */
export const Item = props => {
  const classnames = Item.classnames(props);
  const { data, userDetails, featured } = props;
  const [showOverlay, setShowOverlay] = React.useState(false);
  const { game, ownedBy, recentlyPlayedBy } = data;

  const hasNegativeRating = game.userRating < 50;

  /**
   * returns
   * @param {String} id
   * @param {Array} userDetails
   * @returns {String}
   */
  const getUserAvatar = (id, userDetails) => {
    let avatarPath = "/steam.png";
    for (let i = 0; i < userDetails.length; i++) {
      if (userDetails[i].id === id) {
        avatarPath = userDetails[i].avatarImageUrl;
        break;
      }
    }
    return (
      <img
        className={classnames.element("avatar")}
        key={id}
        src={avatarPath}
        alt={id}
      />
    );
  };

  return (
    <Grid className={classnames.root()} item xs={12} md={featured ? 12 : 6}>
      <Link href="/game/[id]" as={`/game/${game.appid}`}>
        <a className={classnames.element("gamePageLink")}>
          <Paper
            className={classnames.element("paper")}
            onMouseEnter={() => setShowOverlay(true)}
            onMouseLeave={() => setShowOverlay(false)}
          >
            <HeroImage
              altText={game.name}
              imageSrc={`http://steamcdn-a.akamaihd.net/steam/apps/${game.appid}/header.jpg`}
            />
            <div
              className={classnames.element("titleBar", {
                hidden: showOverlay
              })}
            >
              <Typography variant="h1">{game.name}</Typography>
            </div>
            <div
              className={classnames.element("paperOverlayContainer", {
                hidden: !showOverlay
              })}
            >
              <Typography
                className={classnames.element("gameTitle")}
                variant="h3"
              >
                {game.name}
              </Typography>
              <div className={classnames.element("userReviewContainer")}>
                <Typography variant="body2">User Rating:</Typography>
                <Icon
                  path={
                    hasNegativeRating ? mdiThumbDownOutline : mdiThumbUpOutline
                  }
                  className={classnames.element("userReviewIcon", {
                    negative: hasNegativeRating
                  })}
                />
                <Typography variant="body2">{game.userRating}%</Typography>
              </div>
              <Typography variant="body2">
                <span className={classnames.element("categoryTitle")}>
                  Developers:{" "}
                </span>
                {game.developers.join(", ")}
              </Typography>
              <Typography variant="body2">
                <span className={classnames.element("categoryTitle")}>
                  Publishers:{" "}
                </span>
                {game.publishers.join(", ")}
              </Typography>
              <div className={classnames.element("playerContainer")}>
                <div className={classnames.element("playerInfo")}>
                  <Typography
                    className={classnames.element("playerInfoType")}
                    variant="body2"
                  >
                    Owned By:{" "}
                  </Typography>
                  {ownedBy.length === 0 ? (
                    <Typography
                      className={classnames.element("playerInfoTypeEmpty")}
                      variant="body1"
                    >
                      n/a
                    </Typography>
                  ) : (
                    ownedBy.map(id => getUserAvatar(id, userDetails))
                  )}
                </div>
                <div className={classnames.element("playerInfo")}>
                  <Typography
                    className={classnames.element("playerInfoType")}
                    variant="body2"
                  >
                    Recently Played By:{" "}
                  </Typography>
                  {recentlyPlayedBy.length === 0 ? (
                    <Typography
                      className={classnames.element("playerInfoTypeEmpty")}
                      variant="body1"
                    >
                      n/a
                    </Typography>
                  ) : (
                    recentlyPlayedBy.map(id => getUserAvatar(id, userDetails))
                  )}
                </div>
              </div>
              <div className={classnames.element("categoriesContainer")}>
                <Typography variant="body2">
                  <span className={classnames.element("categoryTitle")}>
                    Genres:{" "}
                  </span>
                  {game.genres.join(", ")}
                </Typography>
                <Typography variant="body2">
                  <span className={classnames.element("categoryTitle")}>
                    Tags:{" "}
                  </span>
                  {game.tags.join(", ")}
                </Typography>
              </div>
            </div>
          </Paper>
        </a>
      </Link>
      <div className={classnames.element("priceContainer")}>
        <div
          className={classnames.element("discountPercent", {
            hidden: !game.onSale
          })}
        >
          <Typography variant="h2">-{game.discount}%</Typography>
        </div>
        <div className={classnames.element("prices")}>
          <div
            className={classnames.element("initialPrice", {
              hidden: !game.onSale
            })}
          >
            <Typography color="inherit" variant="body2">
              {formatCurrency(game.initialPrice)}
            </Typography>
          </div>
          <div className={classnames.element("finalPrice")}>
            <Typography variant={game.onSale ? "body2" : "h2"}>
              {game.freeToPlay
                ? "Free to Play"
                : formatCurrency(game.finalPrice)}
            </Typography>
          </div>
        </div>
      </div>
    </Grid>
  );
};
Item.classnames = createClassNameHelper(`${JSS_CLASS_NAME_PREFIX}Item`);
Item.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    gamePageLink: PropTypes.string,
    paper: PropTypes.string,
    titleBar: PropTypes.string,
    titleBarHidden: PropTypes.string,
    gameTitle: PropTypes.string,
    gameImage: PropTypes.string,
    paperOverlayContainer: PropTypes.string,
    paperOverlayContainerHidden: PropTypes.string,
    userReviewContainer: PropTypes.string,
    userReviewIcon: PropTypes.string,
    userReviewIconNegative: PropTypes.string,
    playerContainer: PropTypes.string,
    avatar: PropTypes.string,
    playerInfo: PropTypes.string,
    playerInfoType: PropTypes.string,
    playerInfoTypeEmpty: PropTypes.string,
    categoriesContainer: PropTypes.string,
    categoryTitle: PropTypes.string,
    priceContainer: PropTypes.string,
    discountPercent: PropTypes.string,
    discountPercentHidden: PropTypes.string,
    prices: PropTypes.string,
    initialPrice: PropTypes.string,
    initialPriceHidden: PropTypes.string,
    finalPrice: PropTypes.string
  }),
  // the data to render in the Item
  data: PropTypes.shape({
    game: PropTypes.shape({
      appid: PropTypes.string,
      name: PropTypes.string,
      userRating: PropTypes.number,
      developers: PropTypes.array,
      publishers: PropTypes.array,
      genres: PropTypes.array,
      tags: PropTypes.array,
      onSale: PropTypes.bool,
      freeToPlay: PropTypes.bool,
      discount: PropTypes.number,
      initialPrice: PropTypes.number,
      finalPrice: PropTypes.number
    })
  }).isRequired,
  // profile info for the users
  userDetails: PropTypes.array.isRequired,
  // whether the Item is featured
  featured: PropTypes.bool
};
Item.defaultProps = {
  classes: {}
};

export default withStyles(styles)(Item);
