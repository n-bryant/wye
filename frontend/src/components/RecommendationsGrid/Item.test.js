import { Grid, Typography } from "@material-ui/core/";
import { mdiThumbUpOutline, mdiThumbDownOutline } from "@mdi/js";

import { Item } from "./Item";

describe("Item", () => {
  const props = {
    classes: {
      root: "root",
      gamePageLink: "gamePageLink",
      paper: "paper",
      titleBar: "titleBar",
      titleBarHidden: "titleBarHidden",
      gameTitle: "gameTitle",
      gameImage: "gameImage",
      paperOverlayContainer: "paperOverlayContainer",
      paperOverlayContainerHidden: "paperOverlayContainerHidden",
      userReviewContainer: "userReviewContainer",
      userReviewIcon: "userReviewIcon",
      userReviewIconNegative: "userReviewIconNegative",
      playerContainer: "playerContainer",
      avatar: "avatar",
      playerInfo: "playerInfo",
      playerInfoType: "playerInfoType",
      playerInfoTypeEmpty: "playerInfoTypeEmpty",
      categoriesContainer: "categoriesContainer",
      categoryTitle: "categoryTitle",
      priceContainer: "priceContainer",
      discountPercent: "discountPercent",
      discountPercentHidden: "discountPercentHidden",
      prices: "prices",
      initialPrice: "initialPrice",
      initialPriceHidden: "initialPriceHidden",
      finalPrice: "finalPrice"
    },
    data: {
      game: {
        appid: "1",
        name: "foo",
        userRating: 100,
        developers: ["dev1"],
        publishers: ["pub1", "pub2"],
        genres: ["genre1", "genre2", "genre3"],
        tags: ["tag1", "tag2"],
        onSale: true,
        freeToPlay: false,
        discount: 25,
        initialPrice: 1000,
        finalPrice: 7500
      },
      ownedBy: ["1"],
      recentlyPlayedBy: ["1"]
    },
    userDetails: [{ id: "1", avatarImageUrl: "avatarUrl" }],
    featured: true
  };

  it("should render successfully", () => {
    const wrapper = shallow(<Item {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should set the item size based on whether it is a featured item or not", () => {
    const wrapper = shallow(<Item {...props} />);
    expect(
      wrapper
        .find(Grid)
        .at(0)
        .prop("md")
    ).toBe(10);

    wrapper.setProps({
      ...props,
      featured: false
    });
    wrapper.update();
    expect(
      wrapper
        .find(Grid)
        .at(0)
        .prop("md")
    ).toBe(6);
  });

  it("should show the overlay when hovered", () => {
    const showOverlay = false;
    const setShowOverlay = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [showOverlay, setShowOverlay]);
    const wrapper = shallow(<Item {...props} />);
    wrapper
      .findWhere(n => n.hasClass(props.classes.paper))
      .prop("onMouseEnter")();
    expect(setShowOverlay).toHaveBeenCalledWith(true);
  });

  it("should hide the overlay when no longer hovered", () => {
    const showOverlay = true;
    const setShowOverlay = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [showOverlay, setShowOverlay]);
    const wrapper = shallow(<Item {...props} />);
    wrapper
      .findWhere(n => n.hasClass(props.classes.paper))
      .prop("onMouseLeave")();
    expect(setShowOverlay).toHaveBeenCalledWith(false);
  });

  describe("overlay", () => {
    let wrapper;
    beforeEach(() => {
      const showOverlay = true;
      const setShowOverlay = jest.fn();
      const useStateSpy = jest.spyOn(React, "useState");
      useStateSpy.mockImplementation(() => [showOverlay, setShowOverlay]);
      wrapper = shallow(<Item {...props} />);
    });

    it("should show render a mdiThumbUpOutline if the game has a positive rating", () => {
      expect(
        wrapper
          .findWhere(n => n.hasClass(props.classes.userReviewIcon))
          .prop("path")
      ).toBe(mdiThumbUpOutline);
    });

    it("should render 'n/a' for owned by and recently played by if those values are empty", () => {
      const newProps = {
        classes: {
          root: "root",
          gamePageLink: "gamePageLink",
          paper: "paper",
          titleBar: "titleBar",
          titleBarHidden: "titleBarHidden",
          gameTitle: "gameTitle",
          gameImage: "gameImage",
          paperOverlayContainer: "paperOverlayContainer",
          paperOverlayContainerHidden: "paperOverlayContainerHidden",
          userReviewContainer: "userReviewContainer",
          userReviewIcon: "userReviewIcon",
          userReviewIconNegative: "userReviewIconNegative",
          playerContainer: "playerContainer",
          avatar: "avatar",
          playerInfo: "playerInfo",
          playerInfoType: "playerInfoType",
          playerInfoTypeEmpty: "playerInfoTypeEmpty",
          categoriesContainer: "categoriesContainer",
          categoryTitle: "categoryTitle",
          priceContainer: "priceContainer",
          discountPercent: "discountPercent",
          discountPercentHidden: "discountPercentHidden",
          prices: "prices",
          initialPrice: "initialPrice",
          initialPriceHidden: "initialPriceHidden",
          finalPrice: "finalPrice"
        },
        data: {
          game: {
            appid: "1",
            name: "foo",
            userRating: 4,
            developers: ["dev1"],
            publishers: ["pub1", "pub2"],
            genres: ["genre1", "genre2", "genre3"],
            tags: ["tag1", "tag2"],
            onSale: true,
            freeToPlay: false,
            discount: 25,
            initialPrice: 1000,
            finalPrice: 7500
          },
          ownedBy: ["1"],
          recentlyPlayedBy: ["1"]
        },
        userDetails: [{ id: "1", avatarImageUrl: "avatarUrl" }],
        featured: true
      };
      const newWrapper = shallow(<Item {...newProps} />);
      expect(
        newWrapper
          .findWhere(n => n.hasClass(newProps.classes.userReviewIcon))
          .prop("path")
      ).toBe(mdiThumbDownOutline);
    });

    it("should hide the discount and initial price info if the game is not on sale", () => {
      const newProps = {
        classes: {
          root: "root",
          gamePageLink: "gamePageLink",
          paper: "paper",
          titleBar: "titleBar",
          titleBarHidden: "titleBarHidden",
          gameTitle: "gameTitle",
          gameImage: "gameImage",
          paperOverlayContainer: "paperOverlayContainer",
          paperOverlayContainerHidden: "paperOverlayContainerHidden",
          userReviewContainer: "userReviewContainer",
          userReviewIcon: "userReviewIcon",
          userReviewIconNegative: "userReviewIconNegative",
          playerContainer: "playerContainer",
          avatar: "avatar",
          playerInfo: "playerInfo",
          playerInfoType: "playerInfoType",
          playerInfoTypeEmpty: "playerInfoTypeEmpty",
          categoriesContainer: "categoriesContainer",
          categoryTitle: "categoryTitle",
          priceContainer: "priceContainer",
          discountPercent: "discountPercent",
          discountPercentHidden: "discountPercentHidden",
          prices: "prices",
          initialPrice: "initialPrice",
          initialPriceHidden: "initialPriceHidden",
          finalPrice: "finalPrice"
        },
        data: {
          game: {
            appid: "1",
            name: "foo",
            userRating: 4,
            developers: ["dev1"],
            publishers: ["pub1", "pub2"],
            genres: ["genre1", "genre2", "genre3"],
            tags: ["tag1", "tag2"],
            onSale: false,
            freeToPlay: false,
            discount: 0,
            initialPrice: 7500,
            finalPrice: 7500
          },
          ownedBy: ["1"],
          recentlyPlayedBy: ["1"]
        },
        userDetails: [{ id: "1", avatarImageUrl: "avatarUrl" }],
        featured: true
      };
      const newWrapper = shallow(<Item {...newProps} />);
      expect(
        newWrapper.findWhere(n =>
          n.hasClass(newProps.classes.discountPercentHidden)
        ).length
      ).toBe(1);
      expect(
        newWrapper.findWhere(n =>
          n.hasClass(newProps.classes.initialPriceHidden)
        ).length
      ).toBe(1);
    });

    it("should render 'Free to Play' for the price if the game is free", () => {
      const newProps = {
        classes: {
          root: "root",
          gamePageLink: "gamePageLink",
          paper: "paper",
          titleBar: "titleBar",
          titleBarHidden: "titleBarHidden",
          gameTitle: "gameTitle",
          gameImage: "gameImage",
          paperOverlayContainer: "paperOverlayContainer",
          paperOverlayContainerHidden: "paperOverlayContainerHidden",
          userReviewContainer: "userReviewContainer",
          userReviewIcon: "userReviewIcon",
          userReviewIconNegative: "userReviewIconNegative",
          playerContainer: "playerContainer",
          avatar: "avatar",
          playerInfo: "playerInfo",
          playerInfoType: "playerInfoType",
          playerInfoTypeEmpty: "playerInfoTypeEmpty",
          categoriesContainer: "categoriesContainer",
          categoryTitle: "categoryTitle",
          priceContainer: "priceContainer",
          discountPercent: "discountPercent",
          discountPercentHidden: "discountPercentHidden",
          prices: "prices",
          initialPrice: "initialPrice",
          initialPriceHidden: "initialPriceHidden",
          finalPrice: "finalPrice"
        },
        data: {
          game: {
            appid: "1",
            name: "foo",
            userRating: 4,
            developers: ["dev1"],
            publishers: ["pub1", "pub2"],
            genres: ["genre1", "genre2", "genre3"],
            tags: ["tag1", "tag2"],
            onSale: true,
            freeToPlay: true,
            discount: 0,
            initialPrice: 0,
            finalPrice: 0
          },
          ownedBy: ["1"],
          recentlyPlayedBy: ["1"]
        },
        userDetails: [{ id: "1", avatarImageUrl: "avatarUrl" }],
        featured: true
      };
      const newWrapper = shallow(<Item {...newProps} />);
      expect(
        newWrapper
          .findWhere(n => n.hasClass(newProps.classes.finalPrice))
          .find(Typography)
          .text()
      ).toBe("Free to Play");
    });
  });
});
