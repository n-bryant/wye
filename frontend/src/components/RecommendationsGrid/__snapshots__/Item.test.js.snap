// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item should render successfully 1`] = `
<WithStyles(ForwardRef(Grid))
  className="wye-Item root"
  item={true}
  lg={8}
  md={10}
  xs={12}
>
  <Link
    as="/game/1"
    href="/game/[id]"
  >
    <a
      className="wye-Item__gamePageLink gamePageLink"
    >
      <WithStyles(ForwardRef(Paper))
        className="wye-Item__paper paper"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <WithStyles(HeroImage)
          altText="foo"
          classes={Object {}}
          imageSrc="http://steamcdn-a.akamaihd.net/steam/apps/1/header.jpg"
        />
        <div
          className="wye-Item__titleBar titleBar"
        >
          <WithStyles(ForwardRef(Typography))
            variant="h1"
          >
            foo
          </WithStyles(ForwardRef(Typography))>
        </div>
        <div
          className="wye-Item__paperOverlayContainer wye-Item__paperOverlayContainer--hidden paperOverlayContainer paperOverlayContainerHidden"
        >
          <WithStyles(ForwardRef(Typography))
            className="wye-Item__gameTitle gameTitle"
            variant="h3"
          >
            foo
          </WithStyles(ForwardRef(Typography))>
          <div
            className="wye-Item__userReviewContainer userReviewContainer"
          >
            <WithStyles(ForwardRef(Typography))
              variant="body2"
            >
              User Rating:
            </WithStyles(ForwardRef(Typography))>
            <Icon
              className="wye-Item__userReviewIcon userReviewIcon"
              color={null}
              horizontal={false}
              path="M5,9V21H1V9H5M9,21A2,2 0 0,1 7,19V9C7,8.45 7.22,7.95 7.59,7.59L14.17,1L15.23,2.06C15.5,2.33 15.67,2.7 15.67,3.11L15.64,3.43L14.69,8H21C22.11,8 23,8.9 23,10V12C23,12.26 22.95,12.5 22.86,12.73L19.84,19.78C19.54,20.5 18.83,21 18,21H9M9,19H18.03L21,12V10H12.21L13.34,4.68L9,9.03V19Z"
              rotate={0}
              size={null}
              spin={false}
              vertical={false}
            />
            <WithStyles(ForwardRef(Typography))
              variant="body2"
            >
              100
              %
            </WithStyles(ForwardRef(Typography))>
          </div>
          <WithStyles(ForwardRef(Typography))
            variant="body2"
          >
            <span
              className="wye-Item__categoryTitle categoryTitle"
            >
              Developers:
               
            </span>
            dev1
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Typography))
            variant="body2"
          >
            <span
              className="wye-Item__categoryTitle categoryTitle"
            >
              Publishers:
               
            </span>
            pub1, pub2
          </WithStyles(ForwardRef(Typography))>
          <div
            className="wye-Item__playerContainer playerContainer"
          >
            <div
              className="wye-Item__playerInfo playerInfo"
            >
              <WithStyles(ForwardRef(Typography))
                className="wye-Item__playerInfoType playerInfoType"
                variant="body2"
              >
                Owned By:
                 
              </WithStyles(ForwardRef(Typography))>
              <img
                alt="1"
                className="wye-Item__avatar avatar"
                key="1"
                src="avatarUrl"
              />
            </div>
            <div
              className="wye-Item__playerInfo playerInfo"
            >
              <WithStyles(ForwardRef(Typography))
                className="wye-Item__playerInfoType playerInfoType"
                variant="body2"
              >
                Recently Played By:
                 
              </WithStyles(ForwardRef(Typography))>
              <img
                alt="1"
                className="wye-Item__avatar avatar"
                key="1"
                src="avatarUrl"
              />
            </div>
          </div>
          <div
            className="wye-Item__categoriesContainer categoriesContainer"
          >
            <WithStyles(ForwardRef(Typography))
              variant="body2"
            >
              <span
                className="wye-Item__categoryTitle categoryTitle"
              >
                Genres:
                 
              </span>
              genre1, genre2, genre3
            </WithStyles(ForwardRef(Typography))>
            <WithStyles(ForwardRef(Typography))
              variant="body2"
            >
              <span
                className="wye-Item__categoryTitle categoryTitle"
              >
                Tags:
                 
              </span>
              tag1, tag2
            </WithStyles(ForwardRef(Typography))>
          </div>
        </div>
      </WithStyles(ForwardRef(Paper))>
    </a>
  </Link>
  <div
    className="wye-Item__priceContainer priceContainer"
  >
    <div
      className="wye-Item__discountPercent discountPercent"
    >
      <WithStyles(ForwardRef(Typography))
        variant="h2"
      >
        -
        25
        %
      </WithStyles(ForwardRef(Typography))>
    </div>
    <div
      className="wye-Item__prices prices"
    >
      <div
        className="wye-Item__initialPrice initialPrice"
      >
        <WithStyles(ForwardRef(Typography))
          color="inherit"
          variant="body2"
        >
          $10.00
        </WithStyles(ForwardRef(Typography))>
      </div>
      <div
        className="wye-Item__finalPrice finalPrice"
      >
        <WithStyles(ForwardRef(Typography))
          variant="body2"
        >
          $75.00
        </WithStyles(ForwardRef(Typography))>
      </div>
    </div>
  </div>
</WithStyles(ForwardRef(Grid))>
`;
